***  
### "Genome-wide association of ten agronomic traits in a barley diversity panel"  
June-Sik Kim et al. (under review)  
  
1. ***[211020.pheno-plots.R](https://github.com/junesk9/In-house.codes.published/blob/main/211020.pheno-plots.R)*** â€“ R script for drawing various phenotype plots  
2. ***[211020.qqman-gapit3.R](https://github.com/junesk9/In-house.codes.published/blob/main/211020.qqman-gapit3.R)*** - R script for Manhattan plots from the [GAPIT3](https://github.com/jiabowang/GAPIT3) resultant .GWAS.Results.csv files.  
3. ***[211221.LDHeatmap.R](https://github.com/junesk9/In-house.codes.published/blob/main/211221.LDHeatmap.R)*** - R script to visualize LD blocks with [LDHeatmap](https://sfustatgen.github.io/LDheatmap/).  
4. ***[Density_plot.v2.R](https://github.com/junesk9/In-house.codes.published/blob/main/Density_plot.v2.R)*** - R script to draw transparent density plots.  
5. ***[IBSCv2.EXOME.snp_id.convert.py](https://github.com/junesk9/In-house.codes.published/blob/main/IBSCv2.EXOME.snp_id.convert.py)*** - Python3 script to identify the allocated gene locus for each barley SNP.  
a. The raw Exome-seq sequence data was deposit to the NCBI SRA database [PRJNA825586](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA825586).  
b. The raw exome BeadChip data is available from [NBRP Barley database](http://earth.nig.ac.jp/~dclust/download/iSelect_50K_SV274_A2_DesignStrand.xlsx).   
  

***  
### "Arabidopsis TBP-ASSOCIATED FACTOR 12 ortholog NOBIRO6 controls root elongation with unfolded protein response cofactor activity."
June-Sik Kim et al. (2022) Proc. Natl. Acad. Sci. USA  PMID: [35115407](https://pubmed.ncbi.nlm.nih.gov/35115407/)  

1. ***[200929.box.barplots.scatter.R](https://github.com/junesk9/In-house.codes.published/blob/main/200929.box.barplots.scatter.R)*** - R script for drawing various plots in the manuscript  
2. ***[200208.AT.sleuth-DEG.R](https://github.com/junesk9/In-house.codes.published/blob/main/200208.AT.sleuth-DEG.R)***       - R script for DEG identifying using [Kallsito-Sleuth pipeline](https://www.nature.com/articles/nmeth.4324)  
a. The raw and processed RNA-seq data is available via the NCBI GEO database [GSE163911](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE163911)  
b. The raw DNA-seq data is avaialbel via the NCBI SRA database [PRJNA687636](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA687636)  

***  

